File: leetcode/3-longest-substring/main.cpp
/**
 * PROBLEM DESCRIPTION
 * 3. Longest Substring Without Repeating Characters
 * 
 * Given a string s, find the length of the longest substring without duplicate characters.
 * 
 * Example 1:
 * Input: s = "abcabcbb"
 * Output: 3
 * Explanation: The answer is "abc", with the length of 3.
 */

#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

/**
 * SOLUTION 1:
 * results:
 * runtime: 20ms, beats 26.47%
 * memory: 14.38MB, beats 22.91%
 */
class Solution {
private:
    struct Node {
        Node* next = nullptr;
        char val = 0;
    };

public:
int lengthOfLongestSubstring(string s) {
    unordered_set<char> signs;
    int max_ss = 0;
    int counter = 0;
    for (int it=0; it < s.length(); ++it) {
        while(signs.find(s[it]) != signs.end()) {
            //cout << "found: " << s[it] << " c: " << counter << "\n"; 
            signs.erase(s[counter]);
            counter++;
        }
        signs.insert(s[it]);
        //cout << "add: " << s[it] << " c: " << counter << "\n";
        max_ss = max_ss < (it - counter + 1) ? it - counter + 1 : max_ss;
    }
    return max_ss;
}
};

int main() {
    Solution s;
    string test0 = "abcabcbb";
    string test1 = "bbbbb";
    string test2 = "pwwkew";
    cout << s.lengthOfLongestSubstring(test2) << "\n";
    return 0;
}

File: leetcode/3-longest-substring/main.py
# PROBLEM DESCRIPTION
# 3. Longest Substring Without Repeating Characters
#
# Given a string s, find the length of the longest substring without duplicate characters.
#
# Example 1:
# Input: s = "abcabcbb"
# Output: 3
# Explanation: The answer is "abc", with the length of 3.

# Time Limit Exceeded
# 987 / 988 testcases passed
class Solution1:
    def lengthOfLongestSubstring(self, s: str) -> int:
        substrs = ""
        length = 0
        result = 0
        i = 1
        r = range(0, len(s))
        for i in r:
            if result == len(s):
                break
            for ch in s[i:]:
                if not ch in substrs:
                    substrs += ch
                    length += 1
                    result = length if length > result else result
                else:
                    substrs = ch
                    length = 1
            substrs = ""
            length = 0
        return result

# slicing windows - unique substrings
# going from right if the character was already spotted previously
# we are removing characters from left until we can add this character as unique
# in the worst case we will remove all until this character and start counting new substring
# from beginning, but sometimes only a part of string can be removed to make a substring unique
# and we can continue with counting
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ss = ""
        length = 0
        result = 0
        for ch in s:
            if ch in ss:
                while(ch in ss):
                    ss = ss[1:]
                    length -= 1
            ss += ch
            length += 1
            result = length if length > result else result
        return result
